### DISCUSSION TITLE
Aggregating across included builds

### DISCUSSION THREAD

**Discussion by kelvSYC** [adamko-dev/dokkatoo#211] [Q&A]
I have a project that contains two included builds, which both have the Dokkatoo plugin applied. I wish to create a new included build that generates aggregate docs from the two of them. Setting up the aggregation appears to be simple:

```
includeBuild("../path/to/build1")
includeBuild("../path/to/build2")
```

```
plugins {
    alias(libs.plugins.dokkatoo.html)
}

dependencies {
    dokkatoo("com.mycompany:build1") // included build "build1"
    dokkatoo("com.mycompany:build2") // included build "build2"
}
```

However, the docs generated by the `dokkatooGenerate` does not appear to take the outputs from the `dokkatooGenerateModuleHtml` tasks of the two builds. What else is needed in order for this aggregation to work properly?

**Comment by aSemy** [comment]
Hi 👋

Please make sure each Dokka Module has a distinct module path. The composite build example has a demonstration of how to do this automatically with a convention plugin, check the example's doc for more
https://github.com/adamko-dev/dokkatoo/tree/v2.3.0/examples/composite-build-example/dokkatoo#dokkatoo-composite-build-example


Which version of Dokkatoo are you using? If you're using 2.3.0 then Dokkatoo should log a warning if there's a duplicate module path.

**Reply by kelvSYC** [reply]
Currently using 2.3.0. I've given each module a distinct module path. If I run with `--info`, I did notice something like this:

```
[dokkaHtmlModuleOutputDirectoriesResolver] ignoring artifact html (project :build1) - DokkatooFormatAttribute != html | attributes:{org.gradle.usage=dev.adamko.dokkatoo, dev.adamko.dokkatoo.format=html, dev.adamko.dokkatoo.module-component=ModuleOutputDirectories, artifactType=dokka-module-directory}
```

So I'm suspecting that my declaration of the dependencies is not having any effect.

**Reply by aSemy** [reply]
Hmmm, that's a weird log line. It says that DokkatooFormatAttribute (equivalent to `dev.adamko.dokkatoo.format`) is _not_ `html`, but then in the list of the attributes, `dev.adamko.dokkatoo.format` _is_ `html`. And I don't see that log line in the composite-build-example project.

It's good that there's no warning about duplicate module paths.

I'd like to figure out what the problem is, but that's hard without seeing the config. Is your project open source? Or can you share a reproducer?


What's the structure of the project? Do the included builds have multiple subprojects, or just one root project?

**Reply by kelvSYC** [reply]
The project has two root level builds: `build1` and `build2`, both of which are singular projects, as well as a third included build `aggregation`. `aggregation` has a multi-module project, for which only one module, `dokkatoo`, applies the Dokkatoo plugin. (Other modules in `aggregation` generate other aggregate artifacts and reports.)

This kind of follows the outline used in the Idiomatic Gradle project (https://github.com/jjohannes/idiomatic-gradle), except that in that project, the counterparts to `build1` and `build2` would be one folder level deeper.

**Reply by kelvSYC** [reply]
I've enclosed a small reproducer [dokkatoo-repro.zip](https://github.com/adamko-dev/dokkatoo/files/14880422/dokkatoo-repro.zip)

Not sure if there is a misconfiguration or if there is something going on internally.

**Reply by aSemy** [reply]
Thank you very much! I really appreciate it. I've played around with your project and I can reproduce the problem exactly as you describe. 

I've looked into it, and I've found that when fetching the attribute `variantAttributes[DokkatooFormatAttribute]` returns `null`.

https://github.com/adamko-dev/dokkatoo/blob/4947743f710e75e348ed43c8fb1522c0e7525321/modules/dokkatoo-plugin/src/main/kotlin/dependencies/ModuleComponentDependencies.kt#L106

But then why does the log message show that `dev.adamko.dokkatoo.format=html`???

I'll keep looking into it. I suspect this is a Gradle bug.

EDIT: I've raised a bug on Gradle's side https://github.com/gradle/gradle/issues/28695
